presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -419,11 +419,38 @@ protected GroupedExecutionProperties visitPlan(PlanNode node, Void context)
         @Override
         public GroupedExecutionProperties visitJoin(JoinNode node, Void context)
         {
-            GroupedExecutionProperties properties = processChildren(node);
-            if (properties.isCurrentNodeCapable()) {
-                return new GroupedExecutionProperties(true, true);
+            GroupedExecutionProperties left = node.getLeft().accept(this, null);
+            GroupedExecutionProperties right = node.getRight().accept(this, null);
+
+            if (Optional.empty().equals(node.getDistributionType())) {
+                // This is possible when the optimizers is invoked with `forceSingleNode` set to true.
+                return new GroupedExecutionProperties(false, false);
+            }
+
+            if (Optional.of(JoinNode.DistributionType.REPLICATED).equals(node.getDistributionType())) {
+                // Broadcast join maintains partitioning for the left side.
+                // Right side of a broadcast is not capable of grouped execution because it always comes from a remote exchange.
+                checkState(!right.currentNodeCapable);
```

Comment:
Do i understand correctly that when right side was group-exe-capable, this is cancelled by the fact there are remote exchanges on top?

Reply:
Yes.

This analysis is done after fragmenter. As a result, remote exchange source is a leaf node. Currently, remote exchange doesn't have the capability to have multiple streams for a single recipient task.